### PR TITLE
refactor(idAllocator): refactor id allocator

### PR DESCRIPTION
**Pull request by chunshao90**
# Which issue does this PR close?

Closes #

# Rationale for this change

# What changes are included in this PR?

Refactor ID Allocator with etcdKV.

# Are there any user-facing changes?

no

# How does this change test

Refactor test in server/id/id_test.go

### CODE REVIEW

File: server/id/id_impl.go [hunk 1]
```diff
@@ -5,97 +5,134 @@ package id
...
 
-func (alloc *AllocatorImpl) rebaseLocked(ctx context.Context) error {
-	currEnd, err := alloc.kv.Get(ctx, alloc.key)
+func (a *AllocatorImpl) slowRebaseLocked(ctx context.Context) error {
+	resp, err := a.kv.Get(ctx, a.key)
 	if err != nil {
-		return errors.Wrapf(err, "get end id failed, key:%v", alloc.key)
+		return errors.Wrapf(err, "get end id failed, key:%s", a.key)
+	}
+	var currEnd string
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:70]: Define this variable after error check about the resp.Kvs.

File: server/id/id_impl.go [hunk 2]
```diff
@@ -5,97 +5,134 @@ package id
...
 }
 
-func (alloc *AllocatorImpl) doRebase(ctx context.Context, currEnd uint64) error {
-	newEnd := currEnd + defaultAllocStep
-	key := path.Join(alloc.rootPath, alloc.key)
+func (a *AllocatorImpl) firstDoRebaseLocked(ctx context.Context) error {
+	newEnd := a.allocStep
+
+	cmp := clientv3util.KeyMissing(a.key)
+	op := clientv3.OpPut(a.key, encodeID(newEnd))
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:94]:
```suggestion
	keyMissing := clientv3util.KeyMissing(a.key)
	opPutEnd := clientv3.OpPut(a.key, encodeID(newEnd))
```

File: server/id/id_impl.go [hunk 3]
```diff
@@ -5,97 +5,134 @@ package id
...
+	if currEnd < a.base {
+		return ErrAllocID.WithCausef("ID in storage can't less than memory, base:%d, end:%d", a.base, currEnd)
 	}
-	opPutEndID := clientv3.OpPut(key, encodeID(newEnd))
 
-	resp, err := alloc.kv.Txn(ctx).
+	newEnd := currEnd + a.allocStep
+
+	cmp := clientv3.Compare(clientv3.Value(a.key), "=", encodeID(currEnd))
+	op := clientv3.OpPut(a.key, encodeID(newEnd))
```
Comment (ShiKaiWi) [inline server/id/id_impl.go:120]:
```suggestion
	endEquals := clientv3.Compare(clientv3.Value(a.key), "=", encodeID(currEnd))
	opPutEnd := clientv3.OpPut(a.key, encodeID(newEnd))
```

File: server/id/id_test.go [hunk 4]
```diff
@@ -4,44 +4,72 @@ package id
...
 	})
-	assert.NoError(t, err)
-	rootPath := path.Join("/ceresmeta", strconv.FormatUint(100, 10))
-	kv := storage.NewEtcdKV(client, rootPath)
+	re.NoError(err)
+
+	return client
+}
+
+func TestAllocMultiThread(t *testing.T) {
```
Comment (ShiKaiWi) [inline server/id/id_test.go:40]: - A better name `TestMultipleAllocBasedOnKV`
- No need to allocate id parallelly.

**Review by ShiKaiWi — APPROVED** [review]

**Review by baojinri — APPROVED** [review]
LGTM

**Review by ShiKaiWi — APPROVED** [review]

**Pull request merged**